File: DID.Java.SDK/src/main/java/org/elastos/did/DIDURL.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did;

import java.util.LinkedHashMap;
import java.util.Map;

import org.elastos.did.exception.DIDStoreException;
import org.elastos.did.exception.MalformedDIDURLException;
import org.elastos.did.meta.CredentialMeta;
import org.elastos.did.parser.DIDURLBaseListener;
import org.elastos.did.parser.DIDURLParser;
import org.elastos.did.parser.ParserHelper;

public class DIDURL implements Comparable<DIDURL> {
	private DID did;
	private Map<String, String> parameters;
	private String path;
	private Map<String, String> query;
	private String fragment;

	private CredentialMeta meta;

	public DIDURL(DID id, String fragment) {
		if (id == null)
			throw new IllegalArgumentException();

		if (fragment != null) {
			if (fragment.startsWith("did:")) {
				ParserHelper.parse(fragment, false, new Listener());
				if (!getDid().equals(id))
					throw new IllegalArgumentException("Mismatched arguments");

				return;
			}

			if (fragment.startsWith("#"))
				fragment = fragment.substring(1);
		}

		this.did = id;
		this.fragment = fragment;
	}

	public DIDURL(String url) throws MalformedDIDURLException {
		if (url == null || url.isEmpty())
			throw new IllegalArgumentException();

		try {
			ParserHelper.parse(url, false, new Listener());
		} catch(IllegalArgumentException e) {
			throw new MalformedDIDURLException(e.getMessage());
		}
	}

	public DID getDid() {
		return did;
	}

	public void setDid(DID did) {
		if (did == null)
			throw new IllegalArgumentException();

		this.did = did;
	}

	private String mapToString(Map<String, String> map, String sep) {
		boolean init = true;
		if (map == null)
			return null;

		StringBuilder builder = new StringBuilder(512);
		for (Map.Entry<String, String> entry : map.entrySet()) {
			if (init)
				init = false;
			else
				builder.append(sep);

			builder.append(entry.getKey());
			if (entry.getValue() != null)
				builder.append("=").append(entry.getValue());
		}

		return builder.toString();
	}

	public String getParameters() {
		return mapToString(parameters, ";");
	}

	public String getParameter(String name) {
		if (parameters == null)
			return null;

		return parameters.get(name);
	}

	public boolean hasParameter(String name) {
		if (parameters == null)
			return false;

		return parameters.containsKey(name);
	}

	protected void addParameter(String name, String value) {
		parameters.put(name, value);
	}

	public String getPath() {
		return path;
	}

	protected void setPath(String path) {
		this.path = path;
	}

	public String getQuery() {
		return mapToString(query, "&");
	}

	public String getQueryParameter(String name) {
		if (query == null)
			return null;

		return query.get(name);
	}

	public boolean hasQueryParameter(String name) {
		if (query == null)
			return false;

		return query.containsKey(name);
	}

	protected void addQueryParameter(String name, String value) {
		query.put(name, value);
	}

	public String getFragment() {
		return fragment;
	}

	protected void setFragment(String fragment) {
		this.fragment = fragment;
	}

	protected void setMeta(CredentialMeta meta) {
		this.meta = meta;
	}

	protected CredentialMeta getMeta() {
		if (meta == null)
			meta = new CredentialMeta();

		return meta;
	}

	public void setExtra(String name, String value) throws DIDStoreException {
		if (name == null || name.isEmpty())
			throw new IllegalArgumentException();

		getMeta().setExtra(name, value);

		if (getMeta().attachedStore())
			getMeta().getStore().storeCredentialMeta(this.getDid(), this, meta);
	}

	public String getExtra(String name) {
		if (name == null || name.isEmpty())
			throw new IllegalArgumentException();

		return getMeta().getExtra(name);
	}

	public void setAlias(String alias) throws DIDStoreException {
		getMeta().setAlias(alias);

		if (getMeta().attachedStore())
			if (getMeta().attachedStore())
				getMeta().getStore().storeCredentialMeta(this.getDid(), this, meta);
	}

	public String getAlias() {
		return getMeta().getAlias();
	}
	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder(512);
		builder.append(did);

		if (parameters != null && !parameters.isEmpty())
			builder.append(";").append(getParameters());

		if (path != null && !path.isEmpty())
			builder.append(path);

		if (query != null && !query.isEmpty())
			builder.append("?").append(getQuery());

		if (fragment != null && !fragment.isEmpty())
			builder.append("#").append(getFragment());

		return builder.toString();
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == this)
			return true;

		if (obj instanceof DIDURL) {
			DIDURL id = (DIDURL)obj;
			return toString().equals(id.toString());
		}

		if (obj instanceof String) {
			String url = (String)obj;
			return toString().equals(url);
		}

		return false;
	}

	@Override
	public int compareTo(DIDURL id) {
		return toString().compareTo(id.toString());
	}

	private int mapHashCode(Map<String, String> map) {
		int hash = 0;

		if (map == null)
			return hash;

		for (Map.Entry<String, String> entry : map.entrySet()) {
			hash += entry.getKey().hashCode();
			if (entry.getValue() != null)
				hash += entry.getValue().hashCode();
		}

		return hash;
	}

	@Override
	public int hashCode() {
		int hash = did.hashCode();
		hash += mapHashCode(parameters);
		hash += path == null ? 0 : path.hashCode();
		hash += mapHashCode(query);
		hash += fragment == null ? 0 : fragment.hashCode();

		return hash;
	}

	class Listener extends DIDURLBaseListener {
		private String name;
		private String value;

		@Override
		public void enterDid(DIDURLParser.DidContext ctx) {
			did = new DID();
		}

		@Override
		public void exitMethod(DIDURLParser.MethodContext ctx) {
			String method = ctx.getText();
			if (!method.equals(DID.METHOD))
				throw new IllegalArgumentException("Unknown method: " + method);

			did.setMethod(DID.METHOD);
		}

		@Override
		public void exitMethodSpecificString(
				DIDURLParser.MethodSpecificStringContext ctx) {
			did.setMethodSpecificId(ctx.getText());
		}

		@Override
		public void enterParams(DIDURLParser.ParamsContext ctx) {
			parameters = new LinkedHashMap<String, String>(8);
		}

		@Override
		public void exitParamMethod(DIDURLParser.ParamMethodContext ctx) {
			String method = ctx.getText();
			if (!method.equals(DID.METHOD))
				throw new IllegalArgumentException(
						"Unknown parameter method: " + method);
		}

		@Override
		public void exitParamQName(DIDURLParser.ParamQNameContext ctx) {
			name = ctx.getText();
		}

		@Override
		public void exitParamValue(DIDURLParser.ParamValueContext ctx) {
			value = ctx.getText();
		}

		@Override
		public void exitParam(DIDURLParser.ParamContext ctx) {
			addParameter(name, value);
			name = null;
			value = null;
		}

		@Override
		public void exitPath(DIDURLParser.PathContext ctx) {
			setPath("/" + ctx.getText());
		}

		@Override
		public void enterQuery(DIDURLParser.QueryContext ctx) {
			query = new LinkedHashMap<String, String>(8);
		}

		@Override
		public void exitQueryParamName(DIDURLParser.QueryParamNameContext ctx) {
			name = ctx.getText();
		}

		@Override
		public void exitQueryParamValue(DIDURLParser.QueryParamValueContext ctx) {
			value = ctx.getText();
		}

		@Override
		public void exitQueryParam(DIDURLParser.QueryParamContext ctx) {
			addQueryParameter(name, value);
			name = null;
			value = null;
		}

		@Override
		public void exitFrag(DIDURLParser.FragContext ctx) {
			fragment = ctx.getText();
		}
	}
}

File: DID.Swift.SDK/ElastosDIDSDK/DIDURL.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import Foundation

private func mapToString(_ dict: OrderedDictionary<String, String>?, _ sep: String) -> String? {
    guard let _ = dict else {
        return nil
    }

    var first  = true
    var result = ""

    for (key, value) in dict! {
        result.append(!first ? sep : "")
        result.append(key)
        if value != "" {
            result.append("=")
        }
        result.append(value)

        if  first {
            first = false
        }
    }
    return result
}

/// DID URL defines by the did-url rule, refers to a URL that begins with a DID followed by one or more additional components.
/// A DID URL always identifies the resource to be located. DIDURL includes DID and Url fragment by user defined.
public class DIDURL {
    private static let TAG = "DIDURL"

    private var _did: DID?
    private var _fragment: String?

    private var _parameters: OrderedDictionary<String, String>?
    private var _path: String?
    private var _queryParameters: OrderedDictionary<String, String>?
    private var _metadata: CredentialMeta?

    /// Create a new DID URL according to DID and fragment.
    /// - Parameters:
    ///   - id: A valid didurl guaranteed containing valid did.
    ///   - url: A fragment string.
    /// - Throws: If error occurs, throw error.
    public init(_ id: DID, _ url: String) throws {
        guard !url.isEmpty else {
            throw DIDError.illegalArgument("empty didurl string")
        }

        var fragment = url
        if url.hasPrefix("did:") {
            do {
                try ParserHelper.parse(url, false, DIDURL.Listener(self))
            } catch {
                Log.e(DIDURL.TAG, "Parsing didurl error: malformed didurl string \(url)")
                throw DIDError.malformedDIDURL("malformed DIDURL \(url)")
            }

            guard did == id else {
                throw DIDError.illegalArgument("Mismatched arguments")
            }
            return
        }

        if url.hasPrefix("#") {
            let starIndex = fragment.index(fragment.startIndex, offsetBy: 1)
            let endIndex  = fragment.index(starIndex, offsetBy: fragment.count - 2)
            fragment  = String(fragment[starIndex...endIndex])
        }
        self._did = id
        self._fragment = fragment
    }

    /// Get DID URL from string.
    /// - Parameter url: A  string including id information. idstring support: 1. “did:elastos:xxxxxxx#xxxxx”
    /// - Throws: If error occurs, throw error.
    public init(_ url: String) throws {
        guard !url.isEmpty else {
            throw DIDError.illegalArgument()
        }

        do {
            try ParserHelper.parse(url, false, DIDURL.Listener(self))
        } catch {
            Log.e(DIDURL.TAG, "Parsing didurl error: malformed didurl string \(url)")
            throw DIDError.malformedDIDURL("malformed DIDURL \(url)")
        }
    }

    // A valid didurl guaranteed containing valid did.
    public var did: DID {
        return _did!
    }

    /// Set did
    /// - Parameter newValue: The new did
    public func setDid(_ newValue: DID) {
        self._did = newValue
    }

    // Regards to DIDs v1.0 specs:
    // "DID URL: A DID plus an optional DID path, optional ? character followed
    //  by a DID query, and optional # character followed by a DID fragment."
    public var fragment: String? {
        return _fragment
    }

    func setFragment(_ newValue: String) {
        self._fragment = newValue
    }

    /// <#Description#>
    /// - Returns: <#description#>
    public func parameters() -> String? {
        return mapToString(_parameters, ";")
    }

    public func parameter(ofKey: String) -> String? {
        return _parameters?[ofKey]
    }

    /// Check is contains parameter
    /// - Parameter forKey: <#forKey description#>
    /// - Returns: <#description#>
    public func containsParameter(forKey: String) -> Bool {
        return _parameters?.keys.contains(forKey) ?? false
    }

    func appendParameter(_ value: String?, forKey: String) {
        if  self._parameters == nil {
            self._parameters = OrderedDictionary()
        }
        self._parameters![forKey] = value
    }

    /// <#Description#>
    public var path: String? {
        return _path
    }

    func setPath(_ newValue: String) {
        self._path = newValue
    }

    /// <#Description#>
    /// - Returns: <#description#>
    public func queryParameters() -> String? {
        return mapToString(_queryParameters, "&")
    }

    /// <#Description#>
    /// - Parameter ofKey: <#ofKey description#>
    /// - Returns: <#description#>
    public func queryParameter(ofKey: String) -> String? {
        return _queryParameters?[ofKey]
    }

    /// <#Description#>
    /// - Parameter forKey: <#forKey description#>
    /// - Returns: <#description#>
    public func containsQueryParameter(forKey: String) -> Bool {
        return _queryParameters?.keys.contains(forKey) ?? false
    }

    /// <#Description#>
    /// - Parameters:
    ///   - value: <#value description#>
    ///   - forKey: <#forKey description#>
    public func appendQueryParameter(_ value: String?, forKey: String) {
        if  self._queryParameters == nil {
            self._queryParameters = OrderedDictionary()
        }
        self._queryParameters![forKey] = value
    }

    func setMetadata(_ metadata: CredentialMeta) {
        self._metadata = metadata
    }

    /// Get CredentialMetaData from Credential.
    /// - Returns: Return the handle to CredentialMetaData
    public func getMetadata() -> CredentialMeta {
        if  self._metadata == nil {
            self._metadata = CredentialMeta()
        }
        return self._metadata!
    }

    /// Save Credential(DIDURL) MetaData.
    /// - Throws: If error occurs, throw error.
    public func saveMetadata() throws {
        if (_metadata != nil && _metadata!.attachedStore) {
            try _metadata?.store?.storeCredentialMetadata(did, self, _metadata!)
        }
    }
}

extension DIDURL: CustomStringConvertible {
    func toString() -> String {
        var builder: String = ""

        builder.append(did.toString())
        if (parameters() != nil) {
            builder.append(";")
            builder.append(parameters()!)
        }
        if !(path?.isEmpty ?? true) {
            builder.append(path!)
        }
        if (queryParameters() != nil) {
            builder.append("?")
            builder.append(queryParameters()!)
        }

        if !(_fragment?.isEmpty ?? true) {
            builder.append("#")
            builder.append(fragment!)
        }
        return builder
    }

    /// Get id string from DID URL.
    public var description: String {
        return toString()
    }
}

extension DIDURL: Equatable {
    func equalsTo(_ other: DIDURL) -> Bool {
        return toString() == other.toString()
    }

    func equalsTo(_ other: String) -> Bool {
        return toString() == other
    }

    public static func == (lhs: DIDURL, rhs: DIDURL) -> Bool {
        return lhs.equalsTo(rhs)
    }

    public static func != (lhs: DIDURL, rhs: DIDURL) -> Bool {
        return !lhs.equalsTo(rhs)
    }
}

// DIDURL used as hash key.
extension DIDURL: Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine(self.toString())
    }
}

// Parse Listener
extension DIDURL {
    class Listener: DIDURLBaseListener {
        private var name: String?
        private var value: String?
        private var didURL: DIDURL?

        init(_ didURL: DIDURL) {
            self.didURL = didURL
            super.init()
        }

        override func enterDid(_ ctx: DIDURLParser.DidContext) {
            self.didURL?.setDid(DID())
        }

        override func exitMethod(_ ctx: DIDURLParser.MethodContext) {
            let method = ctx.getText()
            if  method != Constants.METHOD {
                print("Unknown method: \(method)")
            }
            self.didURL?.did.setMethod(Constants.METHOD)
        }

        override func exitMethodSpecificString(
                            _ ctx: DIDURLParser.MethodSpecificStringContext) {
            self.didURL?.did.setMethodSpecificId(ctx.getText())
        }

        override func enterParams(_ ctx: DIDURLParser.ParamsContext) {
            self.didURL?._parameters = OrderedDictionary()
        }

        override func exitParamMethod(_ ctx: DIDURLParser.ParamMethodContext) {
            let method = ctx.getText()
            if  method != Constants.METHOD {
                Log.e(DIDURL.TAG, "Unknown parameter method: \(method)")
            }
            self.didURL?.did.setMethod(method)
        }

        override func exitParamQName(_ ctx: DIDURLParser.ParamQNameContext) {
            self.name = ctx.getText()
        }

        override func exitParamValue(_ ctx: DIDURLParser.ParamValueContext) {
            self.value = ctx.getText()
        }

        override func exitParam(_ ctx: DIDURLParser.ParamContext) {
            let value = self.value ?? ""
            self.didURL?.appendParameter(value, forKey: self.name!)
            self.name = nil
            self.value = nil
        }

        override func exitPath(_ ctx: DIDURLParser.PathContext) {
            self.didURL?.setPath("/" + ctx.getText())
        }

        override func enterQuery(_ ctx: DIDURLParser.QueryContext) {
            self.didURL?._queryParameters = OrderedDictionary()
        }

        override func exitQueryParamName(_ ctx: DIDURLParser.QueryParamNameContext) {
            self.name = ctx.getText()
        }

        override func exitQueryParamValue(_ ctx: DIDURLParser.QueryParamValueContext) {
            self.value = ctx.getText()
        }

        override func exitQueryParam(_ ctx: DIDURLParser.QueryParamContext) {
            let value = self.value ?? ""
            self.didURL?.appendQueryParameter(value, forKey: self.name!)
            self.name = nil
            self.value = nil
        }

        override func exitFrag(_ ctx: DIDURLParser.FragContext) {
            self.didURL?.setFragment(ctx.getText())
        }
    }
}
